lib/algo/manifold_interp/PCHIPInterpolator.cpp

```diff
@@ -0,0 +1,219 @@
+                output_snapshots[counter]->getData()[i] = y_in[j]*computeH1(t,t_in[j],
+                        t_in[j+1]) +
+                        y_in[j+1]*computeH2(t,t_in[j],t_in[j+1]) +
+                        d[j]*computeH3(t,t_in[j],t_in[j+1]) +
+                        d[j+1]*computeH4(t,t_in[j],t_in[j+1]);
+                counter += 1;
+            }
+        }
+
+        d_temp = ((2*h[n_snap-2]+h[n_snap-3])*delta[n_snap-2] - h[n_snap-2]*delta[n_snap
+                  -3])/(h[n_snap-2]+h[n_snap-3]);
+        if(sign(d_temp) != sign(delta[n_snap-2]))
+        {
+            d_temp = 0;
+        }
+        else if (sign(delta[n_snap-2]) != sign(delta[n_snap-3])
+                 && abs(d_temp) > abs(3*delta[n_snap-2]))
+        {
+            d_temp = 3*delta[n_snap-2];
+        }
+        d.push_back(d_temp);
+
+        while(counter < n_out
+                && output_ts[counter]->getData()[0] <= t_in[n_snap-1] + FLT_EPSILON )
+        {
+            t = output_ts[counter]->getData()[0];
+            output_snapshots[counter]->getData()[i] = y_in[n_snap-2]*computeH1(t,
+                    t_in[n_snap-2],t_in[n_snap-1]) +
+                    y_in[n_snap-1]*computeH2(t,t_in[n_snap-2],t_in[n_snap-1]) +
+                    d[n_snap-2]*computeH3(t,t_in[n_snap-2],t_in[n_snap-1]) +
+                    d[n_snap-1]*computeH4(t,t_in[n_snap-2],t_in[n_snap-1]);
+            counter += 1;
+        }
+        CAROM_VERIFY(counter == n_out);
+    }
+}
+
+void PCHIPInterpolator::interpolate(std::vector<Vector*>&
+                                    snapshot_ts,
+                                    std::vector<Vector*>& snapshots,
+                                    int n_out,
+                                    std::vector<Vector*>& output_ts,
+                                    std::vector<Vector*>& output_snapshots)
+{
+    CAROM_VERIFY(snapshot_ts.size() == snapshots.size());
+    CAROM_VERIFY(snapshot_ts.size() > 0);
+    CAROM_VERIFY(n_out > 0);
+    CAROM_VERIFY(output_ts.size() == 0 && output_snapshots.size() == 0);
+
+
+    int n_snap = snapshots.size();
+    int n_dim = snapshots[0]->dim();
+
+
+
+    double t_min = snapshot_ts[0]->getData()[0];
+    double t_max = snapshot_ts[n_snap-1]->getData()[0];
+    double dt = (t_max-t_min)/(n_out-1);
+    output_ts.clear();
+
+    for(int i = 0; i < n_out; ++i)
+    {
+        Vector* temp_t = new Vector(1, false);
+        temp_t->getData()[0] = t_min + i*dt;
+        output_ts.push_back(temp_t);
+    }
+    interpolate(snapshot_ts,snapshots,output_ts, output_snapshots);
+}
+
+double PCHIPInterpolator::computeDerivative(double S1, double S2,
+        double h1,
+        double h2) const
+{
+    double d = 0.0;
+    double alpha = (h1 + 2*h2)/(3*(h1+h2));
+    if(S1*S2 > 0)
+    {
+        d = S1*S2/(alpha*S2 + (1-alpha)*S1);
+    }
+    return d;
+}
+
+double PCHIPInterpolator::computeH1(double x, double xl, double xr) const
+{
+    double h = xr - xl;
```

Comment:
This can be `const double h` for a little more clarity.

Reply:
Changed.